drivers/usb/udc/udc_stm32.c

```diff
@@ -0,0 +1,1128 @@
+
+	status = HAL_PCD_Init(&priv->pcd);
+	if (status != HAL_OK) {
+		LOG_ERR("PCD_Init failed, %d", (int)status);
+		return -EIO;
+	}
+
+	HAL_PCD_Stop(&priv->pcd);
+
+	return 0;
+}
+
+#if defined(USB) || defined(USB_DRD_FS)
+static inline void udc_stm32_mem_init(const struct device *dev)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	const struct udc_stm32_config *cfg = dev->config;
+
+	priv->occupied_mem = cfg->pma_offset;
+}
+
+static int udc_stm32_ep_mem_config(const struct device *dev,
+				   struct udc_ep_config *ep,
+				   bool enable)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	const struct udc_stm32_config *cfg = dev->config;
+	uint32_t size;
+
+	size = MIN(ep->mps, cfg->ep_mps);
+
+	if (!enable) {
+		priv->occupied_mem -= size;
+		return 0;
+	}
+
+	if (priv->occupied_mem + size >= priv->dram_size) {
+		LOG_ERR("Unable to allocate FIFO for 0x%02x", ep->addr);
+		return -ENOMEM;
+	}
+
+	HAL_PCDEx_PMAConfig(&priv->pcd, ep->addr, PCD_SNG_BUF,
+			    priv->occupied_mem);
+
+	priv->occupied_mem += size;
+
+	return 0;
+}
+#else
+static void udc_stm32_mem_init(const struct device *dev)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	const struct udc_stm32_config *cfg = dev->config;
+	int i, words;
+
+	LOG_DBG("DRAM size: %ub", cfg->dram_size);
+
+	if (cfg->ep_mps % 4 || cfg->ep0_mps % 4) {
+		LOG_ERR("Not a 32-bit word multiple: ep0(%u)|ep(%u)",
+			cfg->ep0_mps, cfg->ep_mps);
+		return;
+	}
+
+	/* The documentation is not clear at all about RX FiFo size requirement,
+	 * Allocate a minimum of 0x40 words, which seems to work reliably.
+	 */
+	words = MAX(0x40, cfg->ep_mps / 4);
+	HAL_PCDEx_SetRxFiFo(&priv->pcd, words);
+	priv->occupied_mem = words * 4;
+
+	/* For EP0 TX, reserve only one MPS */
+	HAL_PCDEx_SetTxFiFo(&priv->pcd, 0, cfg->ep0_mps / 4);
+	priv->occupied_mem += cfg->ep0_mps;
+
+	/* Reset TX allocs */
+	for (i = 1U; i < cfg->num_endpoints; i++) {
+		HAL_PCDEx_SetTxFiFo(&priv->pcd, i, 0);
+	}
+}
+
+static int udc_stm32_ep_mem_config(const struct device *dev,
+				   struct udc_ep_config *ep,
+				   bool enable)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	const struct udc_stm32_config *cfg = dev->config;
+	unsigned int words;
+
+	if (!(ep->addr & USB_EP_DIR_IN) || !USB_EP_GET_IDX(ep->addr)) {
+		return 0;
+	}
+
+	if (!enable) {
+		HAL_PCDEx_SetTxFiFo(&priv->pcd, USB_EP_GET_IDX(ep->addr), 0);
+		return 0;
+	}
+
+	words = MIN(ep->mps, cfg->ep_mps) / 4;
+	words = (words <= 64) ? words * 2 : words;
+
+	if (cfg->dram_size - priv->occupied_mem < words * 4) {
+		LOG_ERR("Unable to allocate FIFO for 0x%02x", ep->addr);
+		return -ENOMEM;
+	}
+
+	HAL_PCDEx_SetTxFiFo(&priv->pcd, USB_EP_GET_IDX(ep->addr), words);
+
+	priv->occupied_mem += words * 4;
+
+	return 0;
+}
+#endif
+
+static int udc_stm32_enable(const struct device *dev)
+{
+	struct udc_stm32_data *priv = udc_get_private(dev);
+	const struct udc_stm32_config *cfg = dev->config;
+	HAL_StatusTypeDef status;
+	int ret;
+
+	LOG_DBG("");
```

Comment:
What is the purpose of empty `LOG_DBG` statements?

Reply:
Using that as tracing function debug, as function is prepend to the output, do we want to remove them, is there a cleaner way to do that?